### ISSUE TITLE
treat dependency overrides as third-party

### ISSUE THREAD

**Issue by bholley** [mozilla/cargo-vet#90]
STR:
cd fresh-mozilla-central-clone
cargo vet init
cargo vet suggest

Resulting output: [mio.txt](https://github.com/bholley/cargo-vet/files/8758330/mio.txt)

Note that 'mio' appears several times as a reverse-dependency, but doesn't show up as a recommendation for inspect. We should figure out why that is and resolve the inconsistency.

**Comment by Gankra** [comment]
Mio is being weirdly explicitly patched in here:

https://searchfox.org/mozilla-central/rev/301e9545863b14ff85aa110fabd4274da4593f9d/Cargo.toml#123

As a result in `cargo metadata` it is showing up as a path-dep instead of a crates.io dep (that is vendored). So with the input we get in `vet` has us concluding that it's a first-party crate and not eligible for auditing.

**Comment by Gankra** [comment]
Here's our current first-party graph:


```mermaid
            graph LR
            node8[app_services_logger:0.1.0]
            node15[async-task:4.0.3]
            node21[autocfg:1.0.1]
            node22[baldrdash:0.1.0]
            node26[bench-collections-gtest:0.1.0]
            node35[bits:0.2.0]
            node36[bits_client:0.2.0]
            node37[bitsdownload:0.1.0]
            node42[bookmark_sync:0.1.0]
            node49[cascade_bloom_filter:0.1.0]
            node51[cert_storage:0.0.1]
            node56[chardetng:0.1.9]
            node57[chardetng_c:0.1.2]
            node72[coremidi:0.6.0]
            node92[d3d12:0.4.1]
            node99[defaultagent-static:0.1.0]
            node101[derive_common:0.0.1]
            node115[encoding_glue:0.1.0]
            node120[error-support:0.1.0]
            node133[fluent-ffi:0.1.0]
            node135[fluent-langneg-ffi:0.1.0]
            node139[fog:0.1.0]
            node140[fog-gtest:0.1.0]
            node141[fog_control:0.1.0]
            node157[gecko-profiler:0.1.0]
            node158[gecko_logger:0.1.0]
            node159[geckodriver:0.31.0]
            node160[geckoservo:0.0.1]
            node163[gkrust:0.1.0]
            node164[gkrust-gtest:0.1.0]
            node165[gkrust-shared:0.1.0]
            node166[gkrust_utils:0.1.0]
            node175[glsl-to-cxx:0.1.0]
            node177[gluesmith:0.1.0]
            node179[golden_gate:0.1.0]
            node194[http3server:0.1.1]
            node195[http_sfv:0.1.0]
            node206[interrupt-support:0.1.0]
            node210[ipcclientcerts-static:0.1.0]
            node215[js-sys:0.3.100]
            node216[jsparagus:0.1.0]
            node217[jsparagus-ast:0.1.0]
            node218[jsparagus-emitter:0.1.0]
            node219[jsparagus-generated-parser:0.1.0]
            node220[jsparagus-json-log:0.1.0]
            node221[jsparagus-parser:0.1.0]
            node222[jsparagus-scope:0.1.0]
            node223[jsparagus-stencil:0.1.0]
            node224[jsrust:0.1.0]
            node225[jsrust_shared:0.1.0]
            node228[kvstore:0.1.0]
            node229[l10nregistry:0.3.0]
            node230[l10nregistry-ffi:0.1.0]
            node231[l10nregistry-ffi-gtest:0.1.0]
            node241[libudev-sys:0.1.3]
            node246[localization-ffi:0.1.0]
            node251[malloc_size_of:0.0.1]
            node253[mapped_hyph:0.4.3]
            node254[marionette:0.2.0]
            node256[mdns_service:0.1.0]
            node263[metal:0.23.1]
            node264[midir:0.7.0]
            node265[midir_impl:0.1.0]
            node268[minidump_writer_linux:0.1.0]
            node270[mio:0.6.23]
            node273[moz_asserts:0.1.0]
            node275[moz_task:0.1.0]
            node276[moz_task-gtest:0.1.0]
            node277[mozbuild:0.1.0]
            node278[mozdevice:0.5.0]
            node279[mozglue-static:0.1.0]
            node280[mozilla-central-workspace-hack:0.1.0]
            node281[mozprofile:0.8.0]
            node282[mozrunner:0.14.0]
            node283[mozurl:0.0.1]
            node284[mozversion:0.4.3]
            node285[mozwer_s:0.1.0]
            node286[mp4parse:0.13.0]
            node287[mp4parse-gtest:0.1.0]
            node288[mp4parse_capi:0.13.0]
            node291[naga:0.8.0]
            node292[neqo-common:0.5.7]
            node293[neqo-crypto:0.5.7]
            node294[neqo-http3:0.5.7]
            node295[neqo-qpack:0.5.7]
            node296[neqo-transport:0.5.7]
            node297[neqo_glue:0.1.0]
            node299[netwerk_helper:0.0.1]
            node303[nserror:0.1.0]
            node304[nss_build_common:0.1.0]
            node305[nsstring:0.1.0]
            node306[nsstring-gtest:0.1.0]
            node318[origin-trials-ffi:0.1.0]
            node320[osclientcerts-static:0.1.4]
            node327[peek-poke:0.2.0]
            node328[peek-poke-derive:0.2.1]
            node349[prefs_parser:0.0.1]
            node352[processtools:0.1.0]
            node353[profiler-macros:0.1.0]
            node354[profiler_helper:0.1.0]
            node358[qcms:0.2.0]
            node375[regex-ffi:0.1.0]
            node384[rsclientcerts:0.1.0]
            node385[rsdparsa_capi:0.1.0]
            node392[rust_minidump_writer_linux:0.1.0]
            node403[selectors:0.22.0]
            node418[servo_arc:0.1.1]
            node424[size_of_test:0.0.1]
            node426[slotmap:1.100.0]
            node429[smoosh:0.1.0]
            node432[sql-support:0.1.0]
            node435[static_prefs:0.1.0]
            node436[storage:0.1.0]
            node437[storage_variant:0.1.0]
            node439[style:0.0.1]
            node440[style_derive:0.0.1]
            node441[style_traits:0.0.1]
            node442[stylo_tests:0.0.1]
            node444[swgl:0.1.0]
            node446[sync-guid:0.1.0]
            node447[sync15-traits:0.1.0]
            node460[to_shmem:0.0.1]
            node461[to_shmem_derive:0.0.1]
            node472[unic-langid-ffi:0.1.0]
            node487[vcpkg:0.2.999]
            node489[viaduct:0.1.0]
            node495[wasm-bindgen:0.2.100]
            node501[web-sys:0.3.100]
            node502[webdriver:0.45.0]
            node503[webext-storage:0.1.0]
            node504[webext_storage_bridge:0.1.0]
            node505[webrender:0.61.0]
            node506[webrender_api:0.61.0]
            node507[webrender_bindings:0.1.0]
            node508[webrender_build:0.0.1]
            node510[wgpu-core:0.12.0]
            node511[wgpu-hal:0.12.0]
            node512[wgpu-types:0.12.0]
            node513[wgpu_bindings:0.1.0]
            node519[wineventlog:0.1.0]
            node522[wr_malloc_size_of:0.0.1]
            node525[xpcom:0.1.0]
            node526[xpcom-gtest:0.1.0]
            node527[xpcom_macros:0.1.0]
            node528[xulstore:0.1.0]
            node8 --> node179
            node8 --> node303
            node8 --> node305
            node8 --> node525
            node22 --> node277
            node36 --> node35
            node37 --> node36
            node37 --> node275
            node37 --> node303
            node37 --> node305
            node37 --> node525
            node42 --> node275
            node42 --> node303
            node42 --> node305
            node42 --> node436
            node42 --> node437
            node42 --> node525
            node49 --> node303
            node49 --> node305
            node49 --> node525
            node51 --> node275
            node51 --> node303
            node51 --> node305
            node51 --> node437
            node51 --> node522
            node51 --> node525
            node57 --> node56
            node99 --> node280
            node99 --> node489
            node99 --> node519
            node115 --> node303
            node115 --> node305
            node133 --> node229
            node133 --> node305
            node133 --> node525
            node135 --> node303
            node135 --> node305
            node135 --> node472
            node135 --> node525
            node139 --> node277
            node140 --> node139
            node141 --> node139
            node141 --> node303
            node141 --> node305
            node141 --> node435
            node141 --> node489
            node141 --> node525
            node157 --> node277
            node157 --> node353
            node158 --> node8
            node159 --> node254
            node159 --> node278
            node159 --> node281
            node159 --> node282
            node159 --> node284
            node159 --> node502
            node160 --> node157
            node160 --> node251
            node160 --> node305
            node160 --> node403
            node160 --> node418
            node160 --> node439
            node160 --> node441
            node160 --> node460
            node163 --> node165
            node163 --> node279
            node163 --> node280
            node163 --> node442
            node163 --> node444
            node164 --> node26
            node164 --> node140
            node164 --> node165
            node164 --> node231
            node164 --> node276
            node164 --> node279
            node164 --> node287
            node164 --> node306
            node164 --> node444
            node164 --> node526
            node165 --> node8
            node165 --> node49
            node165 --> node51
            node165 --> node57
            node165 --> node115
            node165 --> node133
            node165 --> node135
            node165 --> node141
            node165 --> node157
            node165 --> node158
            node165 --> node160
            node165 --> node166
            node165 --> node195
            node165 --> node225
            node165 --> node228
            node165 --> node229
            node165 --> node230
            node165 --> node246
            node165 --> node253
            node165 --> node273
            node165 --> node283
            node165 --> node288
            node165 --> node297
            node165 --> node299
            node165 --> node303
            node165 --> node305
            node165 --> node318
            node165 --> node349
            node165 --> node352
            node165 --> node358
            node165 --> node375
            node165 --> node385
            node165 --> node435
            node165 --> node436
            node165 --> node472
            node165 --> node489
            node165 --> node504
            node165 --> node507
            node165 --> node513
            node165 --> node525
            node166 --> node305
            node179 --> node206
            node179 --> node275
            node179 --> node303
            node179 --> node305
            node179 --> node437
            node179 --> node447
            node179 --> node525
            node194 --> node270
            node194 --> node292
            node194 --> node293
            node194 --> node294
            node194 --> node295
            node194 --> node296
            node195 --> node303
            node195 --> node305
            node195 --> node525
            node210 --> node384
            node216 --> node217
            node216 --> node218
            node216 --> node219
            node216 --> node220
            node216 --> node221
            node216 --> node222
            node216 --> node223
            node218 --> node217
            node218 --> node222
            node218 --> node223
            node219 --> node217
            node221 --> node217
            node221 --> node219
            node221 --> node220
            node222 --> node217
            node222 --> node223
            node223 --> node217
            node224 --> node225
            node224 --> node279
            node225 --> node279
            node225 --> node280
            node228 --> node275
            node228 --> node303
            node228 --> node305
            node228 --> node437
            node228 --> node525
            node230 --> node133
            node230 --> node229
            node230 --> node275
            node230 --> node303
            node230 --> node305
            node230 --> node525
            node231 --> node230
            node231 --> node275
            node246 --> node133
            node246 --> node229
            node246 --> node230
            node246 --> node275
            node246 --> node303
            node246 --> node305
            node246 --> node525
            node251 --> node403
            node251 --> node418
            node256 --> node157
            node264 --> node72
            node264 --> node215
            node264 --> node495
            node264 --> node501
            node265 --> node264
            node265 --> node305
            node273 --> node277
            node275 --> node15
            node275 --> node303
            node275 --> node305
            node275 --> node525
            node276 --> node275
            node279 --> node277
            node282 --> node281
            node283 --> node303
            node283 --> node305
            node283 --> node525
            node285 --> node280
            node288 --> node286
            node293 --> node277
            node293 --> node292
            node294 --> node292
            node294 --> node293
            node294 --> node295
            node294 --> node296
            node295 --> node292
            node295 --> node293
            node295 --> node296
            node296 --> node292
            node296 --> node293
            node297 --> node292
            node297 --> node293
            node297 --> node294
            node297 --> node295
            node297 --> node296
            node297 --> node303
            node297 --> node305
            node297 --> node525
            node299 --> node303
            node299 --> node305
            node303 --> node277
            node303 --> node305
            node306 --> node305
            node320 --> node280
            node320 --> node384
            node327 --> node328
            node352 --> node303
            node352 --> node525
            node375 --> node305
            node385 --> node303
            node392 --> node268
            node392 --> node305
            node403 --> node418
            node403 --> node460
            node403 --> node461
            node429 --> node216
            node432 --> node206
            node432 --> node304
            node435 --> node277
            node436 --> node303
            node436 --> node305
            node436 --> node437
            node436 --> node525
            node437 --> node303
            node437 --> node305
            node437 --> node525
            node439 --> node157
            node439 --> node251
            node439 --> node277
            node439 --> node305
            node439 --> node403
            node439 --> node418
            node439 --> node435
            node439 --> node440
            node439 --> node441
            node439 --> node460
            node439 --> node461
            node440 --> node101
            node441 --> node251
            node441 --> node403
            node441 --> node418
            node441 --> node460
            node441 --> node461
            node442 --> node160
            node442 --> node251
            node442 --> node403
            node442 --> node424
            node442 --> node439
            node442 --> node441
            node442 --> node460
            node444 --> node175
            node444 --> node508
            node447 --> node446
            node460 --> node418
            node461 --> node101
            node472 --> node303
            node472 --> node305
            node472 --> node525
            node503 --> node120
            node503 --> node206
            node503 --> node304
            node503 --> node432
            node503 --> node446
            node503 --> node447
            node504 --> node179
            node504 --> node275
            node504 --> node303
            node504 --> node305
            node504 --> node432
            node504 --> node437
            node504 --> node503
            node504 --> node525
            node505 --> node444
            node505 --> node506
            node505 --> node508
            node505 --> node522
            node506 --> node327
            node506 --> node522
            node507 --> node157
            node507 --> node305
            node507 --> node444
            node507 --> node505
            node507 --> node522
            node510 --> node291
            node510 --> node511
            node510 --> node512
            node511 --> node92
            node511 --> node215
            node511 --> node263
            node511 --> node291
            node511 --> node495
            node511 --> node501
            node511 --> node512
            node513 --> node510
            node513 --> node511
            node513 --> node512
            node525 --> node277
            node525 --> node303
            node525 --> node305
            node525 --> node527
            node526 --> node303
            node526 --> node305
            node526 --> node525
            node527 --> node277
            node528 --> node275
            node528 --> node303
            node528 --> node305
            node528 --> node525
 ```

**Comment by Gankra** [comment]
atlas: 

* diamonds are "roots"
* parallelograms are "workspace members"
* rectangles are "other first-parties"

**Comment by Gankra** [comment]
Ah nice, I think subgraphs make this cleaner:

```mermaid
graph LR
    subgraph roots
        node22{baldrdash:0.1.0}
        node37{bitsdownload:0.1.0}
        node42{bookmark_sync:0.1.0}
        node99{defaultagent-static:0.1.0}
        node159{geckodriver:0.31.0}
        node163{gkrust:0.1.0}
        node164{gkrust-gtest:0.1.0}
        node177{gluesmith:0.1.0}
        node194{http3server:0.1.1}
        node210{ipcclientcerts-static:0.1.0}
        node224{jsrust:0.1.0}
        node256{mdns_service:0.1.0}
        node265{midir_impl:0.1.0}
        node285{mozwer_s:0.1.0}
        node320{osclientcerts-static:0.1.4}
        node354{profiler_helper:0.1.0}
        node392{rust_minidump_writer_linux:0.1.0}
        node429{smoosh:0.1.0}
        node528{xulstore:0.1.0}
    end
    subgraph workspace-members
        node8[/app_services_logger:0.1.0/]
        node26[/bench-collections-gtest:0.1.0/]
        node35[/bits:0.2.0/]
        node36[/bits_client:0.2.0/]
        node49[/cascade_bloom_filter:0.1.0/]
        node51[/cert_storage:0.0.1/]
        node115[/encoding_glue:0.1.0/]
        node133[/fluent-ffi:0.1.0/]
        node135[/fluent-langneg-ffi:0.1.0/]
        node139[/fog:0.1.0/]
        node140[/fog-gtest:0.1.0/]
        node141[/fog_control:0.1.0/]
        node157[/gecko-profiler:0.1.0/]
        node158[/gecko_logger:0.1.0/]
        node165[/gkrust-shared:0.1.0/]
        node179[/golden_gate:0.1.0/]
        node195[/http_sfv:0.1.0/]
        node225[/jsrust_shared:0.1.0/]
        node228[/kvstore:0.1.0/]
        node229[/l10nregistry:0.3.0/]
        node230[/l10nregistry-ffi:0.1.0/]
        node231[/l10nregistry-ffi-gtest:0.1.0/]
        node246[/localization-ffi:0.1.0/]
        node254[/marionette:0.2.0/]
        node275[/moz_task:0.1.0/]
        node276[/moz_task-gtest:0.1.0/]
        node278[/mozdevice:0.5.0/]
        node279[/mozglue-static:0.1.0/]
        node280[/mozilla-central-workspace-hack:0.1.0/]
        node281[/mozprofile:0.8.0/]
        node282[/mozrunner:0.14.0/]
        node283[/mozurl:0.0.1/]
        node284[/mozversion:0.4.3/]
        node287[/mp4parse-gtest:0.1.0/]
        node297[/neqo_glue:0.1.0/]
        node299[/netwerk_helper:0.0.1/]
        node303[/nserror:0.1.0/]
        node305[/nsstring:0.1.0/]
        node306[/nsstring-gtest:0.1.0/]
        node349[/prefs_parser:0.0.1/]
        node352[/processtools:0.1.0/]
        node353[/profiler-macros:0.1.0/]
        node358[/qcms:0.2.0/]
        node375[/regex-ffi:0.1.0/]
        node384[/rsclientcerts:0.1.0/]
        node385[/rsdparsa_capi:0.1.0/]
        node435[/static_prefs:0.1.0/]
        node436[/storage:0.1.0/]
        node437[/storage_variant:0.1.0/]
        node472[/unic-langid-ffi:0.1.0/]
        node502[/webdriver:0.45.0/]
        node504[/webext_storage_bridge:0.1.0/]
        node513[/wgpu_bindings:0.1.0/]
        node519[/wineventlog:0.1.0/]
        node525[/xpcom:0.1.0/]
        node526[/xpcom-gtest:0.1.0/]
        node527[/xpcom_macros:0.1.0/]
    end
    subgraph first-party
        node15[async-task:4.0.3]
        node21[autocfg:1.0.1]
        node56[chardetng:0.1.9]
        node57[chardetng_c:0.1.2]
        node72[coremidi:0.6.0]
        node92[d3d12:0.4.1]
        node101[derive_common:0.0.1]
        node120[error-support:0.1.0]
        node160[geckoservo:0.0.1]
        node166[gkrust_utils:0.1.0]
        node175[glsl-to-cxx:0.1.0]
        node206[interrupt-support:0.1.0]
        node215[js-sys:0.3.100]
        node216[jsparagus:0.1.0]
        node217[jsparagus-ast:0.1.0]
        node218[jsparagus-emitter:0.1.0]
        node219[jsparagus-generated-parser:0.1.0]
        node220[jsparagus-json-log:0.1.0]
        node221[jsparagus-parser:0.1.0]
        node222[jsparagus-scope:0.1.0]
        node223[jsparagus-stencil:0.1.0]
        node241[libudev-sys:0.1.3]
        node251[malloc_size_of:0.0.1]
        node253[mapped_hyph:0.4.3]
        node263[metal:0.23.1]
        node264[midir:0.7.0]
        node268[minidump_writer_linux:0.1.0]
        node270[mio:0.6.23]
        node273[moz_asserts:0.1.0]
        node277[mozbuild:0.1.0]
        node286[mp4parse:0.13.0]
        node288[mp4parse_capi:0.13.0]
        node291[naga:0.8.0]
        node292[neqo-common:0.5.7]
        node293[neqo-crypto:0.5.7]
        node294[neqo-http3:0.5.7]
        node295[neqo-qpack:0.5.7]
        node296[neqo-transport:0.5.7]
        node304[nss_build_common:0.1.0]
        node318[origin-trials-ffi:0.1.0]
        node327[peek-poke:0.2.0]
        node328[peek-poke-derive:0.2.1]
        node403[selectors:0.22.0]
        node418[servo_arc:0.1.1]
        node424[size_of_test:0.0.1]
        node426[slotmap:1.100.0]
        node432[sql-support:0.1.0]
        node439[style:0.0.1]
        node440[style_derive:0.0.1]
        node441[style_traits:0.0.1]
        node442[stylo_tests:0.0.1]
        node444[swgl:0.1.0]
        node446[sync-guid:0.1.0]
        node447[sync15-traits:0.1.0]
        node460[to_shmem:0.0.1]
        node461[to_shmem_derive:0.0.1]
        node487[vcpkg:0.2.999]
        node489[viaduct:0.1.0]
        node495[wasm-bindgen:0.2.100]
        node501[web-sys:0.3.100]
        node503[webext-storage:0.1.0]
        node505[webrender:0.61.0]
        node506[webrender_api:0.61.0]
        node507[webrender_bindings:0.1.0]
        node508[webrender_build:0.0.1]
        node510[wgpu-core:0.12.0]
        node511[wgpu-hal:0.12.0]
        node512[wgpu-types:0.12.0]
        node522[wr_malloc_size_of:0.0.1]
    end
    subgraph third-party
    end
    node8 --> node179
    node8 --> node303
    node8 --> node305
    node8 --> node525
    node22 --> node277
    node36 --> node35
    node37 --> node36
    node37 --> node275
    node37 --> node303
    node37 --> node305
    node37 --> node525
    node42 --> node275
    node42 --> node303
    node42 --> node305
    node42 --> node436
    node42 --> node437
    node42 --> node525
    node49 --> node303
    node49 --> node305
    node49 --> node525
    node51 --> node275
    node51 --> node303
    node51 --> node305
    node51 --> node437
    node51 --> node522
    node51 --> node525
    node57 --> node56
    node99 --> node280
    node99 --> node489
    node99 --> node519
    node115 --> node303
    node115 --> node305
    node133 --> node229
    node133 --> node305
    node133 --> node525
    node135 --> node303
    node135 --> node305
    node135 --> node472
    node135 --> node525
    node139 --> node277
    node140 --> node139
    node141 --> node139
    node141 --> node303
    node141 --> node305
    node141 --> node435
    node141 --> node489
    node141 --> node525
    node157 --> node277
    node157 --> node353
    node158 --> node8
    node159 --> node254
    node159 --> node278
    node159 --> node281
    node159 --> node282
    node159 --> node284
    node159 --> node502
    node160 --> node157
    node160 --> node251
    node160 --> node305
    node160 --> node403
    node160 --> node418
    node160 --> node439
    node160 --> node441
    node160 --> node460
    node163 --> node165
    node163 --> node279
    node163 --> node280
    node163 --> node442
    node163 --> node444
    node164 --> node26
    node164 --> node140
    node164 --> node165
    node164 --> node231
    node164 --> node276
    node164 --> node279
    node164 --> node287
    node164 --> node306
    node164 --> node444
    node164 --> node526
    node165 --> node8
    node165 --> node49
    node165 --> node51
    node165 --> node57
    node165 --> node115
    node165 --> node133
    node165 --> node135
    node165 --> node141
    node165 --> node157
    node165 --> node158
    node165 --> node160
    node165 --> node166
    node165 --> node195
    node165 --> node225
    node165 --> node228
    node165 --> node229
    node165 --> node230
    node165 --> node246
    node165 --> node253
    node165 --> node273
    node165 --> node283
    node165 --> node288
    node165 --> node297
    node165 --> node299
    node165 --> node303
    node165 --> node305
    node165 --> node318
    node165 --> node349
    node165 --> node352
    node165 --> node358
    node165 --> node375
    node165 --> node385
    node165 --> node435
    node165 --> node436
    node165 --> node472
    node165 --> node489
    node165 --> node504
    node165 --> node507
    node165 --> node513
    node165 --> node525
    node166 --> node305
    node179 --> node206
    node179 --> node275
    node179 --> node303
    node179 --> node305
    node179 --> node437
    node179 --> node447
    node179 --> node525
    node194 --> node270
    node194 --> node292
    node194 --> node293
    node194 --> node294
    node194 --> node295
    node194 --> node296
    node195 --> node303
    node195 --> node305
    node195 --> node525
    node210 --> node384
    node216 --> node217
    node216 --> node218
    node216 --> node219
    node216 --> node220
    node216 --> node221
    node216 --> node222
    node216 --> node223
    node218 --> node217
    node218 --> node222
    node218 --> node223
    node219 --> node217
    node221 --> node217
    node221 --> node219
    node221 --> node220
    node222 --> node217
    node222 --> node223
    node223 --> node217
    node224 --> node225
    node224 --> node279
    node225 --> node279
    node225 --> node280
    node228 --> node275
    node228 --> node303
    node228 --> node305
    node228 --> node437
    node228 --> node525
    node230 --> node133
    node230 --> node229
    node230 --> node275
    node230 --> node303
    node230 --> node305
    node230 --> node525
    node231 --> node230
    node231 --> node275
    node246 --> node133
    node246 --> node229
    node246 --> node230
    node246 --> node275
    node246 --> node303
    node246 --> node305
    node246 --> node525
    node251 --> node403
    node251 --> node418
    node256 --> node157
    node264 --> node72
    node264 --> node215
    node264 --> node495
    node264 --> node501
    node265 --> node264
    node265 --> node305
    node273 --> node277
    node275 --> node15
    node275 --> node303
    node275 --> node305
    node275 --> node525
    node276 --> node275
    node279 --> node277
    node282 --> node281
    node283 --> node303
    node283 --> node305
    node283 --> node525
    node285 --> node280
    node288 --> node286
    node293 --> node277
    node293 --> node292
    node294 --> node292
    node294 --> node293
    node294 --> node295
    node294 --> node296
    node295 --> node292
    node295 --> node293
    node295 --> node296
    node296 --> node292
    node296 --> node293
    node297 --> node292
    node297 --> node293
    node297 --> node294
    node297 --> node295
    node297 --> node296
    node297 --> node303
    node297 --> node305
    node297 --> node525
    node299 --> node303
    node299 --> node305
    node303 --> node277
    node303 --> node305
    node306 --> node305
    node320 --> node280
    node320 --> node384
    node327 --> node328
    node352 --> node303
    node352 --> node525
    node375 --> node305
    node385 --> node303
    node392 --> node268
    node392 --> node305
    node403 --> node418
    node403 --> node460
    node403 --> node461
    node429 --> node216
    node432 --> node206
    node432 --> node304
    node435 --> node277
    node436 --> node303
    node436 --> node305
    node436 --> node437
    node436 --> node525
    node437 --> node303
    node437 --> node305
    node437 --> node525
    node439 --> node157
    node439 --> node251
    node439 --> node277
    node439 --> node305
    node439 --> node403
    node439 --> node418
    node439 --> node435
    node439 --> node440
    node439 --> node441
    node439 --> node460
    node439 --> node461
    node440 --> node101
    node441 --> node251
    node441 --> node403
    node441 --> node418
    node441 --> node460
    node441 --> node461
    node442 --> node160
    node442 --> node251
    node442 --> node403
    node442 --> node424
    node442 --> node439
    node442 --> node441
    node442 --> node460
    node444 --> node175
    node444 --> node508
    node447 --> node446
    node460 --> node418
    node461 --> node101
    node472 --> node303
    node472 --> node305
    node472 --> node525
    node503 --> node120
    node503 --> node206
    node503 --> node304
    node503 --> node432
    node503 --> node446
    node503 --> node447
    node504 --> node179
    node504 --> node275
    node504 --> node303
    node504 --> node305
    node504 --> node432
    node504 --> node437
    node504 --> node503
    node504 --> node525
    node505 --> node444
    node505 --> node506
    node505 --> node508
    node505 --> node522
    node506 --> node327
    node506 --> node522
    node507 --> node157
    node507 --> node305
    node507 --> node444
    node507 --> node505
    node507 --> node522
    node510 --> node291
    node510 --> node511
    node510 --> node512
    node511 --> node92
    node511 --> node215
    node511 --> node263
    node511 --> node291
    node511 --> node495
    node511 --> node501
    node511 --> node512
    node513 --> node510
    node513 --> node511
    node513 --> node512
    node525 --> node277
    node525 --> node303
    node525 --> node305
    node525 --> node527
    node526 --> node303
    node526 --> node305
    node526 --> node525
    node527 --> node277
    node528 --> node275
    node528 --> node303
    node528 --> node305
    node528 --> node525
```

**Comment by Gankra** [comment]
This is using #92 which I hacked up this morning because I wanted some more debugging tools.

**Comment by Gankra** [comment]
Status: unclear how to do this, it doesn't appear that cargo-metadata lets you "know" that a `[patch]` has happened (but I should look closer to confirm). One possible option is to check if a path/git dep has a `(name, version)` pair that exists on crates.io, and if it does to assume it's really third-party?

However the docs you wrote for this seem a bit confused.

> The reason is that this custom version might be generated in one of two ways: by building a semantically-compatible replacement from scratch, or by starting with the source of the original crate and making some (potentially-minimal) modifications. The latter case is quite common, and in practice rarely entails a full audit of the original crate despite formally transforming it into first-party code. Since cargo vet has no way to distinguish this case from a from-scratch rewrite, it conservatively assumes the former. The latter case can be handled by adding an entry to the unaudited table with suggest = false and a note explaining the situation.

This seems to be saying that a from-scratch-rewrite is the case that merits an audit, but that's the case that seems "obviously truly first-party" and therefore needing no audits. The "small patch to the base crate" case on the other hand does seem like it wants a proper audit chain.

However either case is problematic to include in your audits because **other people can import your audits and don't have the context that your audit of `serde:1.0.0` is actually your fork and may differ from what they have in-tree**.

**Comment by bholley** [comment]
> Status: unclear how to do this, it doesn't appear that cargo-metadata lets you "know" that a `[patch]` has happened (but I should look closer to confirm). One possible option is to check if a path/git dep has a `(name, version)` pair that exists on crates.io, and if it does to assume it's really third-party?

IIRC the plan was to use "has third-party reverse-dependencies" as a heuristic for now and then submit a fix upstream to have `cargo metadata` expose this info.

> However the docs you wrote for this seem a bit confused.
> 
> > The reason is that this custom version might be generated in one of two ways: by building a semantically-compatible replacement from scratch, or by starting with the source of the original crate and making some (potentially-minimal) modifications. The latter case is quite common, and in practice rarely entails a full audit of the original crate despite formally transforming it into first-party code. Since cargo vet has no way to distinguish this case from a from-scratch rewrite, it conservatively assumes the former. The latter case can be handled by adding an entry to the unaudited table with suggest = false and a note explaining the situation.
> 
> This seems to be saying that a from-scratch-rewrite is the case that merits an audit, but that's the case that seems "obviously truly first-party" and therefore needing no audits. The "small patch to the base crate" case on the other hand does seem like it wants a proper audit chain.

Hah! The confusion here is just because the options are enumerated twice, in different orders, making "former" ambiguous. I'll fix.

> However either case is problematic to include in your audits because **other people can import your audits and don't have the context that your audit of `serde:1.0.0` is actually your fork and may differ from what they have in-tree**.

The expectation is that people would audit the original version, not any fork they have (that's certainly what `cargo vet inspect` would do). I can update the docs to make that more explicit.

**Comment by bholley** [comment]
Nika made the good point that people [sometimes](https://searchfox.org/mozilla-central/rev/552bfc6334b797d92fb2ce8e93a6ace5f47cd56d/Cargo.toml#159-164) use `[patch]` just to achieve simple first-party referencing of non-public crates. So even if `cargo metadata` were to tell us what was patched, we couldn't confidently assume that such crates mapped to anything on crates.io.

So I think we'll have to live with "has third-party reverse-dependencies", plus perhaps a `config.toml` directive to force something to require audits.

**Comment by bholley** [comment]
OK, we discussed at length and settled on the following:

Policy tables get a new field called `audit-as-crates-io` which is `Option<bool>` and defaults to `None`. If `--locked` is not passed, `cargo vet` checks all first-party crates (name, version) against the registry, and requires any matches to have a `Some` entry. We also update the documentation accordingly.

**Comment by mystor** [comment]
I believe this should've been fixed as part of #173
